### ISSUE TITLE
word: true doesn't appear to work with "^"

### ISSUE THREAD

**Issue by JPUnD** [federico-terzi/espanso#345]
Hi there,

I've made a list of subscripts and superscripts that i could find - [all of them here](https://github.com/JPUnD/espanso-super-and-sub-scripts) - and there seems to be some of them that don't work correctly.

I'm using several triggers, as an example here's two of them:
  - trigger: ":^0"
    replace: "⁰"
    word: true

  - trigger: ":_0"
    replace: "₀"
    word: true

Typing ":_0" on the keyboard with a space after it triggers ₀ as expected however typing ":^0" with a space afterward does not trigger ⁰ as expected.

I assume that there's something going wrong with the "^" symbol because taking off the `word = true` statement still doesn't fix the ":^0" not being triggered. 

If i could make a suggestion most if not all of the times i'm using this subscripts and superscripts is to use in variable names when discussing mathematical equations so it would be nice to have a scenario where instead of "word" you would have something like "half-word" so that something like "K:_a" would trigger to "Kₐ" which doesn't because `word: true` only works with a space before and after the word, and i would like the trigger to act even if there's something behind the word.

**Comment by federico-terzi** [comment]
Hey,

> I've made a list of subscripts and superscripts that i could find - [all of them here](https://github.com/JPUnD/espanso-super-and-sub-scripts) - and there seems to be some of them that don't work correctly.
> 
> I'm using several triggers, as an example here's two of them:
> 
> * trigger: ":^0"
>   replace: "⁰"
>   word: true
> * trigger: ":_0"
>   replace: "₀"
>   word: true
> 
> Typing ":_0" on the keyboard with a space after it triggers ₀ as expected however typing ":^0" with a space afterward does not trigger ⁰ as expected.
> 
> I assume that there's something going wrong with the "^" symbol because taking off the `word = true` statement still doesn't fix the ":^0" not being triggered.

That's interesting, thanks for pointing that out. Which OS and keyboard layout are you using?

> If i could make a suggestion most if not all of the times i'm using this subscripts and superscripts is to use in variable names when discussing mathematical equations so it would be nice to have a scenario where instead of "word" you would have something like "half-word" so that something like "K:_a" would trigger to "Kₐ" which doesn't because `word: true` only works with a space before and after the word, and i would like the trigger to act even if there's something behind the word.

What about adding a space to the end of the trigger? Such as `":_0 "` and without the `word` option. This should work pretty much as you want, with the only exception being that it only triggers if the following character is a space (and if you want something else, you could just press SPACE -> BACKSPACE -> OTHER CHAR).

Cheers :)

**Comment by JPUnD** [comment]
I'm running Manjaro on kernel 4.19.126, using i3 as a WM on (Portuguese qwerty keyboard)[https://upload.wikimedia.org/wikipedia/commons/thumb/2/2c/KB_Portuguese.svg/1200px-KB_Portuguese.svg.png]. Espanso was installed from the AUR package espanso-bin 0.6.2-1.

When it comes to the last suggestion that does seem to be a good workaround, but that was just a problem because triggers like ":^1" and ":^1/2" were on conflict for obvious reasons, but i think ill just get rid of ":^1" since my keyboard already gives me the symbol ¹ without the need for espanso.

**Comment by federico-terzi** [comment]
All right, thank you @JPUnD. As soon as I get my hands on a Linux machine I'll check

**Comment by jpmvferreira** [comment]
I've tested this again but instead of having the trigger to be `^` I've also tried `~` and `´`, and it didn't work, which means that it might be due to a problem using combining characters because replacing the`^` with `u` worked as expected.

**Comment by federico-terzi** [comment]
@jpmvferreira Interesting, thank you. I'm pretty sure this has to do with the Portuguese layout. Perhaps the new version might help, we'll see :)

**Comment by SpecialCharacter** [comment]
I have German keyboard layout under Windows 10 (have not checked under Mac yet) with the same issue.
^a --> â etc. works because it is translated to another character, integrating the deadkey. ^z is not translated into ẑ because the German keyboard layout is not equipped for that, so the ^ just goes from invisible to visible character.
Is there a possibility to "read out" a deadkey already in the invisible state?

**Comment by federico-terzi** [comment]
@SpecialCharacter Interesting, thank you... Is there any chance you could post a screen recording of both of them happening? (both the a and z case)

That would really help :)

**Comment by SpecialCharacter** [comment]
Sorry, never done that ...
trigger: "â" / replace: ":)"--> output ":)"
trigger: "^a" / replace: ":)" --> output "â" [German keyboard layout]
trigger: "ẑ" / replace: ":("--> output "^z" [Would need Esperanto keyboard layout for that, I think]
trigger: "^z" / replace: ":(" --> output "^z"

**Comment by federico-terzi** [comment]
@SpecialCharacter I see, thank you! I'll investigate

**Comment by SpecialCharacter** [comment]
Any update?

**Comment by federico-terzi** [comment]
No updates on this one, I had to focus my energies on the #594 release, after that we'll move on lower-priority issues :)

**Comment by Oskiator** [comment]
Hi ! 
I read all the above and I think I have a similar issue, but I'm not sure I've understood everything properly.

In my case, I'm on win 10, switzerland french keyboard layout.

I tried several triggers using "^".

What I saw : 
When a character string starts with â, ê, ô etc. (for example "â" alone or "âxyz". The trigger works properly.
For example, "ê" get replaced by "être".

But when the character string contain a â, ê, ô... that is not at the beginning, the trigger doesn't work,
Example, trigger "tâ", replace "tâche" doesn't work.

**Comment by SpecialCharacter** [comment]
Oskiator, what about axyz (without â)? Does it trigger as well?
Or "tâ" --> "tache"?
Maybe it strips the Unicode and works just in ASCII...?

**Comment by federico-terzi** [comment]
@Oskiator What version of Espanso are you running? You can find out by launching `espanso run` in a terminal

**Comment by Oskiator** [comment]
@SpecialCharacter 

1) replace works well with â, ê... It's just with the triggers that I have some problems. 
So axyz trigger works fine, also when the word to replace contain "â".
e.g : axyz --> "tâche" works. axyz --> "têche" also works. 

2) When there is nothing before the â AND "word: false" (or not specified), it works. 

**Works**       
  - trigger: "ât" 
    replace: "tâche"

  - trigger: "â" 
    replace: "tâche"

**Doesn't work**
  - trigger: "tâ" 
    replace: "tâche"

  - trigger: "â" 
    replace: "tâche"
    word: true
  
  - trigger: "ta" 
    replace: "tâche"
    word: true      
    
  - trigger: "ât" 
    replace: "tâche"
    word: true   

@federico-terzi 
"espanso run" didn't work for me but "espanso --help" or "espanso --version" worked.

=> espanso 2.1.4-beta

**Comment by federico-terzi** [comment]
@Oskiator Thanks! We'll investigate

> "espanso run" didn't work for me but "espanso --help" or "espanso --version" worked.

You're absolutely right! My bad

**Comment by James-Mat** [comment]
I'm having the same problem; also german keyboard layout and trying to expand to superscripts, 2.1.5-beta, Windows 10.0.19044. A fix for that would be really nice for mathematical expansions!

There's also a crash happening in conjunction with combining characters, but since this issue here isn't really about that, I opened a separate one for that (#1157).

@SpecialCharacter for me the trigger `^z` works, if I don't type to the dead character; what I mean is that you can force a deadkey by typing space afterwards, so typing `^`,` `,`z` results in `^z` which for me gets substituted correctly; however, typing `^`,`z`, which, since the german typeset has no ẑ, results in `^z` as well, doesn't get substituted.

@jpmvferreira The same workaround works for "^0", "^1", etc.: pressing the space key in between to force the ^ to appear by itself makes the trigger function as expected. I'll probably be using this until the combination without space works as well.  
If espanso crashes for you on undo backspace as well, I'd suggest adding `undo_backspace: false` to the default.yml though if the crash mentioned in #1157 happens for you as well.

Undo backspace can't be disabled for individual matches, can it?

**Comment by jpmvferreira** [comment]
For me adding a space doesn't between the `^` and the character that I want to specify as an upper doesn't work.

**Comment by smeech** [comment]
This is now quite an old issue. I am unable to reproduce the problem on a UK keyboard, as all the triggers work as expected.
Is it still a problem?

**Comment by jpmvferreira** [comment]
Yes, at least to me, I've just tried it out once again in my Portuguese keyboard and using `^` as  a trigger does not work.
Espanso version is 2.2.1, installed from the package `espanso-bin` from the AUR.

**Comment by smeech** [comment]
Thank you. We now have a number of similar issues which appear to involve non-English keyboard layouts, which we'll need to pull together somehow.

**Comment by smeech** [comment]
Also:
https://github.com/espanso/espanso/issues/498
https://github.com/espanso/espanso/issues/511
https://github.com/espanso/espanso/issues/996
https://github.com/espanso/espanso/issues/1112
https://github.com/espanso/espanso/issues/1133
https://github.com/espanso/espanso/issues/1286
https://github.com/espanso/espanso/issues/1153
https://github.com/espanso/espanso/issues/1609
